File: src/SimpleAnonymizer.java
import java.io.File;
import java.util.ArrayList;
import java.util.List;

import com.sun.javaws.jnl.LaunchSelection;
import com.swabunga.spell.engine.SpellDictionary;
import com.swabunga.spell.engine.SpellDictionaryHashMap;
import com.swabunga.spell.engine.Word;
import com.swabunga.spell.event.SpellCheckEvent;
import com.swabunga.spell.event.SpellCheckListener;
import com.swabunga.spell.event.SpellChecker;
import com.swabunga.spell.event.StringWordTokenizer;
import edu.washington.cs.knowitall.morpha.MorphaStemmer;


public class SimpleAnonymizer implements SpellCheckListener{

    protected Boolean testing = true;

    protected ArrayList<String> _nicknames;
    protected SpellChecker spellCheck = null;
    protected SpellDictionary dictionary = null;
    protected ArrayList<String> misspelled;


    private String AnonymizeWord(String W, String FirstNameR, String LastNameR, String FirstNameO, String LastNameO, Boolean F) {
        String result = W;

        result = AnonymizeWord_Pronoun(result);
        result = AnonymizeWord_Adjective(result);
        result = AnonymizeWord_Abbreviations(result);
        result = AnonymizeWord_Name(W, FirstNameR, LastNameR);
        result = AnonymizeWord_Name(W, FirstNameO, LastNameO);
        result = AnonymizeWord_NickName(W, FirstNameR, LastNameR);
        result = AnonymizeWord_NickName(W, FirstNameO, LastNameO);
        result = AnonymizeWord_NameSubword(W, FirstNameR);
        result = AnonymizeWord_NameSubword(W, LastNameR);
        result = AnonymizeWord_NameSubword(W, FirstNameO);
        result = AnonymizeWord_NameSubword(W, LastNameO);

        result = AnonmizeWord_Flag(W, F);

        return result;
    }

    private String AnonymizeWord_Adjective(String W){
        String result = W;

        if (W.equals("male")  ||
            W.equals("female")
        )
        {
            result = "ADJECTIVE";
        }

        return result;
    }

    private String AnonymizeWord_Abbreviations(String W){
        String result = W;

        if (W.equals("m")  ||
            W.equals("f")
        )
        {
            result = "FLAGGED";
        }

        return result;
    }

    private String AnonymizeWord_Pronoun(String W){
        String result = W;

        if (W.equals("he")  ||
            W.equals("she") ||
            W.equals("him") ||
            W.equals("her") ||
            W.equals("his") ||
            W.equals("hers") ||
            W.equals("himself") ||
            W.equals("herself") ||
            W.equals("man") ||
            W.equals("woman") ||
            W.equals("men") ||
            W.equals("women") ||
            W.equals("boy") ||
            W.equals("girl") ||
            W.equals("mother") ||
            W.equals("father") ||
            W.equals("lady") ||
            W.equals("uncle") ||
            W.equals("aunt") ||
            W.equals("son") ||
            W.equals("daughter") ||
            W.equals("husband") ||
            W.equals("wife")
        )
        {
            result = "PRONOUN";
        }

        return result;
    }

    private String AnonymizeWord_Name(String W, String FirstName, String LastName){
        String result = W;

        if (W.equals(FirstName)){
            result = "FIRSTNAME";
        }
        else if (W.equals(LastName)){
            result = "LASTNAME";
        }

        return result;
    }

    private String AnonymizeWord_NickName(String W, String FirstName, String LastName){
        String result = W;

        // should make come from a database probably

        if (_nicknames.contains(FirstName))
        {
            result = "NICKNAME";
        }
        else if (_nicknames.contains(LastName))
        {
            result = "NICKNAME";
        }

        return result;
    }

    private String AnonymizeWord_NameSubword(String W, String Name){
        String result = W;
        String processing = W;

        Boolean flagDone = false; // indicates if processing should end
        Boolean flagPrefix = false; // track if the prefix matches a possible name/nickname
        Boolean flagSpelling = false; // tracks if the word is misspelled
        Boolean flagNotWord = false; // tracks if the word is not recognized at all even as an English word
        Boolean flagSuffix = false; // tracks if the suffix is in general odd

        // does the word end in Y, I, K, etc.?
        Boolean flagSuffixY = false;
        Boolean flagSuffixI = false;
        Boolean flagSuffixK = false;
        Boolean flagSuffixH = false;
        Boolean flagSuffixIE = false;
        Boolean flagSuffixCH = false;
        Boolean flagSuffixKY = false;
        Boolean flagSuffixHY = false;

        misspelled = new ArrayList<String>();
        spellCheck.checkSpelling(new StringWordTokenizer(W));

        if (misspelled.size() > 0) {
            for (int iError = 0; iError < misspelled.size(); iError++) {
                List<Word> suggestions = spellCheck.getSuggestions(misspelled.get(iError), 1);
                if (suggestions.size() == 0) {
                    flagNotWord = true;
                }
                else{
                    flagSpelling = true;
                }
            }
        }

        // if the name starts with the word it is not a word
        if (Name.startsWith(W) && flagNotWord){
            result = "NICKNAME?";
            flagPrefix = true;
            flagDone = true;
        }

        // if the name starts with the word
        if (Name.startsWith(W)){
            result = "FLAGGED";
            flagPrefix = true;
            flagDone = true;
        }

        if(!flagDone){
            // does the word end in y or i or ie
            // if so strip it off
            if (processing.endsWith("y")){
                flagSuffixY = true;
                flagSuffix = true;
                processing = processing.substring(0,processing.length()-1);
            }
            else if (processing.endsWith("i")){
                flagSuffixI = true;
                flagSuffix = true;
                processing = processing.substring(0,processing.length()-1);
            }
            else if (processing.endsWith("k")){
                flagSuffixK = true;
                flagSuffix = true;
                processing = processing.substring(0,processing.length()-1);
            }
            else if (processing.endsWith("h")){
                flagSuffixH = true;
                flagSuffix = true;
                processing = processing.substring(0,processing.length()-1);
            }

            if (W.startsWith(processing)){
                result = "NICKNAME?";
                flagPrefix = true;
            }

            // if this is not already flagged, then take a look at strange two letter endings
            // reset the word
            processing = W;
            if (!flagPrefix){
                if (processing.endsWith("ie")){
                    flagSuffixIE = true;
                    flagSuffix = true;
                    processing = processing.substring(0,processing.length()-2);
                }
                else if (processing.endsWith("ch")){
                    flagSuffixCH = true;
                    flagSuffix = true;
                    processing = processing.substring(0,processing.length()-2);
                }
                else if (processing.endsWith("ky")){
                    flagSuffixKY = true;
                    flagSuffix = true;
                    processing = processing.substring(0,processing.length()-2);
                }
                else if (processing.endsWith("hy")){
                    flagSuffixHY = true;
                    flagSuffix = true;
                    processing = processing.substring(0,processing.length()-2);
                }

                // if after removing the ending the word matches the
                if (W.startsWith(processing)){
                    result = "NICKNAME?";
                    flagPrefix = true;
                }
            }

            // this word has a strange suffix, flag it
            if (!flagPrefix && (flagSuffixIE || flagSuffixI || flagSuffixKY)){
                result = "FLAGGED";
            }

            // this isn't a word and it has a strange ending
            if (flagNotWord && flagSuffix){
                result = "FLAGGED";
            }
        }

        return result;
    }

    private String AnonmizeWord_Flag(String W, Boolean F){
        String result = W;

        if (F){
            result = "FLAGGED";
        }

        return result;
    }



    private Boolean SetAnonFlag(String W){
        Boolean result = false;

        if (W.equals("doctor") ||
            W.equals("dr") ||
            W.equals("mister") ||
            W.equals("miss") ||
            W.equals("mr") ||
            W.equals("ms") ||
            W.equals("mrs")
           ){
            return true;
        }

        return result;
    }

    public String Anonymize(ArrayList<String> I){
        String output = "";
        ArrayList<String> sample = I;

        String resident_name = I.get(0);
        String observer_name = I.get(1);
        String feedback = I.get(2);
        String professionalism_safety = I.get(3);

        String[] nameTokens = resident_name.split("\\s");
        String firstnameR = nameTokens[0].toLowerCase();
        String lastnameR = nameTokens[nameTokens.length-1].toLowerCase();

        nameTokens = observer_name.split("\\s");
        String firstnameO = nameTokens[0].toLowerCase();
        String lastnameO = nameTokens[nameTokens.length-1].toLowerCase();

        System.out.println(resident_name + " = " + firstnameR + ", " + lastnameR);
        System.out.println(observer_name + " = " + firstnameO + ", " + lastnameO);

        String[] feedbackTokens = feedback.split("\\s");
        String[] prosafeTokens = professionalism_safety.split("\\s");

        ArrayList<String> anonFeedback = new ArrayList<>();
        ArrayList<String> anonProSafe = new ArrayList<>();

        Boolean flagAnon = false;

        for (int iToken = 0; iToken < feedbackTokens.length; iToken++){
            feedbackTokens[iToken] = feedbackTokens[iToken].replaceAll("[^\\p{IsAlphabetic}\\p{IsDigit}]", "");
            feedbackTokens[iToken] = new MorphaStemmer().morpha(feedbackTokens[iToken], false);
            anonFeedback.add(AnonymizeWord(feedbackTokens[iToken], firstnameR, lastnameR, firstnameO, lastnameO, flagAnon));
            flagAnon = SetAnonFlag(feedbackTokens[iToken]);
        }

        for (int iToken = 0; iToken < prosafeTokens.length; iToken++){
            prosafeTokens[iToken] = prosafeTokens[iToken].replaceAll("[^\\p{IsAlphabetic}\\p{IsDigit}]", "");
            prosafeTokens[iToken] = new MorphaStemmer().morpha(prosafeTokens[iToken], false);
            anonProSafe.add(AnonymizeWord(prosafeTokens[iToken], firstnameR, lastnameR, firstnameO, lastnameO, flagAnon));
            flagAnon = SetAnonFlag(prosafeTokens[iToken]);
        }

        System.out.println("Feedback");
        System.out.println("========");

        for (int iToken = 0; iToken < feedbackTokens.length; iToken++){
            System.out.println(feedbackTokens[iToken] + " -> " + anonFeedback.get(iToken));
        }

        System.out.println("");
        System.out.println("Professional & Safety");
        System.out.println("=====================");

        for (int iToken = 0; iToken < prosafeTokens.length; iToken++){
            System.out.println(prosafeTokens[iToken] + " -> " + anonProSafe.get(iToken));
        }

        return output;
    }

    protected void InitializeDictionary()
    {
        // initialize spell checking
        try {
            //String dictFileName = "./dict/eng_com.dic";
            String dictFileName = "./dict/english.0";
            String phonetFileName = "./dict/phonet.en";

            File dictFile = new File(dictFileName);
            File phonetFile = new File(phonetFileName);

            dictionary = new SpellDictionaryHashMap(dictFile, phonetFile);
            spellCheck = new SpellChecker(dictionary);
            spellCheck.addSpellCheckListener(this);

            // add all words in the master answer key to the dictionary
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    protected boolean ValidateDictionary()
    {// currently requires no validation
        boolean result = true;
        return result;
    }

    public void spellingError(SpellCheckEvent event)
    {
        event.ignoreWord(true);
        misspelled.add(event.getInvalidWord());
    }


    public SimpleAnonymizer(ArrayList<String> Nicknames)
    {
        _nicknames = Nicknames;

        InitializeDictionary();
        try{
            if (!ValidateDictionary())
            {
                throw new Exception("Invalid dictionary");
            }
        }
        catch(Exception e){
            System.out.println(e.getMessage());
        }
    }

    private void Test(){
        // ensure that all test cases are lowercase
    }
}

File: SimpleAnonymizer.py
# import spellchecker
# from spellchecker import SpellChecker

_nicknames = ["Rick", "Ricky", "Richie", "Dick"]

"""
Inputs:
Takes a word W

Returns:
Either the word W or an anonymized word

Function:
Works by check to see if W is a known pronoun

@todo The list should not really be hardcoded but should come from some external source?
"""


def AnonymizeWord_Pronoun(W):
    result = W

    if ((W == "he") or
            (W == "she") or
            (W == "him") or
            (W == "her") or
            (W == "his") or
            (W == "hers") or
            (W == "himself") or
            (W == "herself") or
            (W == "man") or
            (W == "woman") or
            (W == "men") or
            (W == "women") or
            (W == "boy") or
            (W == "girl") or
            (W == "mother") or
            (W == "father") or
            (W == "lady") or
            (W == "uncle") or
            (W == "aunt") or
            (W == "son") or
            (W == "daughter") or
            (W == "husband") or
            (W == "wife")):
        result = "PRONOUN"

    return result


"""
Inputs:
    W: A string. The word under consideration.

Returns:
    result: A string. Either the original word or the anonymized word

Function:
    Checks to see if the word is a known adjective that needs to be anonymized.
    Currently, this is "male" and "female".

@todo The list should not really be hardcoded but should come from some external source?    
"""


def AnonymizeWord_Adjective(W):
    result = W

    if ((W == "male") or
            (W == "female")):
        result = "ADJECTIVE"

    return result


"""
Inputs:
    W: A string. The word under consideration.

Returns:
    result: A string. Either the original word or the anonymized word

Function:
    Checks to see if the word is a known abbreviation.
    Currently, this is "m" (for male) and "f" (for female). However, other abbreviations may be added.

@todo The list should not really be hardcoded but should come from some external source?    
"""


def AnonymizeWord_Abbreviations(W):
    result = W

    if ((W == "m") or
            (W == "f")):
        result = "FLAGGED"

    return result


"""
Inputs:
    W: The word under consideration
    Name: A name of a person associated to the EPA record. Usually this will be either the resident or observer
    Note the name could be the first, last, or middle name

Returns:
    result: Either the original word W or the anonymized word

Function:
    Simply checks to see if the word is exactly the same as the first name or last name
"""


def AnonymizeWord_Name(W, Name):
    result = W

    if (W == Name):
        result = "NAME"

    return result


"""
Inputs:
    W: The word under consideration
    Name: A name of a person associated to the EPA record. Usually this will be either the resident or observer
    Note the name could be the first, last, or middle name

Returns:
    result: Either the original word W or the anonymized word

Function:
    Checks to see if the name is a known nickname stored in the global variable _nicknames (this was a property in the Java version)
    @todo Currently, this is a hardcoded list but it really should come from an external source
    @todo Also,it might be necessary to have nickname associated to a name. E.g., Ricky associated to Richard 
"""


def AnonymizeWord_NickName(W, Name):
    result = W
    found = False
    index = 0

    while (index < len(_nicknames)) and (not found):
        # seperated the check in case we want to return a different value for first and last name nicknames
        if (Name == _nicknames[index].lower()):
            result = "NICKNAME"
            found = True
        index = index + 1

    return result


"""
Inputs:
    W: A string. This is the word under consideration.
    Name: A string. This can be a first or last name.

Returns:
    result: A string. This is either the original word or the anonymized word

Functions:
    This function attempts to determine if a word is a nickname (or name) by applying a series of logical rules
    Each of the rules usually will set a flag, and then different combinations of flags are checked at the end.

    @todo We need a spellchecker for Python
"""


def AnonymizeWord_NameSubword(W, Name):
    result = W
    processing = W  # as the word is modified, this is used to prevent changing the original

    flagDone = False  # indicates if functioning should terminate
    flagPrefix = False  # track if the prefix matches a possible name / nickname
    flagSpelling = False  # tracks if the word is misspelled (currently not working, need a spellchecker)
    flagNotWord = False  # tracks if the word is not recognized  at all (currently not working, need a spellchecker)
    flagSuffix = False  # tracks if the  suffix is in general odd

    # does the word end in Y, I, K, etc.?
    flagSuffixY = False
    flagSuffixI = False
    flagSuffixK = False
    flagSuffixH = False
    flagSuffixIE = False
    flagSuffixCH = False
    flagSuffixKY = False
    flagSuffixHY = False

    # Need a Python based spellchecker to replace this Java code
    # pyspellchecker is too good, it recognizes names properly
    #    spell = SpellChecker()

    # find those words that may be misspelled
    #   misspelled = spell.unknown(['brent'])
    #   for word in misspelled:
    #      print(word + ": " + str(spell.candidates(word)))

    # if the name starts with the W, and W is not a word
    # currently not working, need spellchecker
    if (Name.find(W) == 0 and flagNotWord):
        result = "NICKNAME?"
        flagPrefix = True
        flagDone = True

    # if the name starts with the W
    if (Name.find(W) == 0):
        result = "FLAGGED"
        flagPrefix = True
        flagDone = True

    if (not flagDone):
        # does the word end in y or i or ie
        # if so strip it off
        if (processing.endswith("y")):
            flagSuffixY = True;
            flagSuffix = True;
            processing = processing[0:len(processing) - 1]
        elif (processing.endswith("i")):
            flagSuffixI = True
            flagSuffix = True;
            processing = processing[0:len(processing) - 1]
        elif (processing.endswith("k")):
            flagSuffixK = True
            flagSuffix = True
            processing = processing[0:len(processing) - 1]
        elif (processing.endswith("h")):
            flagSuffixH = True
            flagSuffix = True
            processing = processing[0:len(processing) - 1]

        # check to see if the original word has the processing string as a prefix
        # if so it is a likely nickname
        if (flagSuffix and W.startswith(processing)):
            result = "NICKNAME?"
            flagPrefix = True

        # if this is not already flagged, then take a look at strange two letter endings
        # reset the processing string
        processing = W
        if (not flagPrefix):
            if (processing.endswith("ie")):
                flagSuffixIE = True
                flagSuffix = True
                processing = processing[0:len(processing) - 2]
            elif (processing.endswith("ch")):
                flagSuffixCH = True
                flagSuffix = True
                processing = processing[0:len(processing) - 2]
            elif (processing.endswith("ky")):
                flagSuffixKY = True
                flagSuffix = True
                processing = processing[0:len(processing) - 2]
            elif (processing.endswith("hy")):
                flagSuffixHY = True
                flagSuffix = True
                processing = processing[0:len(processing) - 2]

            # if after removing the ending the word matches the
            if (flagSuffix and W.startswith(processing)):
                result = "NICKNAME?"
                flagPrefix = True

    # this word has a strange suffix, flag it
    if (not flagPrefix and (flagSuffixIE or flagSuffixI or flagSuffixKY)):
        result = "FLAGGED"

    # this isn't a word and it has a strange ending
    if (flagNotWord and flagSuffix):
        result = "FLAGGED"

    return result


def AnonymizeWord_Flag(W, F):
    result = W

    if (F):
        result = "FLAGGED"

    return result


"""
DEPRECATED - Replaced with a new version that takes an array of names
Inputs:
    String W: The current word under consideration
    String FirstNameR: The resident's first name
    String LastNameR: The resident's last name
    String FirstNameO: The observer's first name
    String LastNameO: The observer's last name
    Boolean F: A flag that indicates the previous word was unusual

Returns:
    String: the word W or the anonymized word

AnonymizeWord functions by calling several other functions to do the actual work of anonymization.
Think of it as a collection of rules to be applied.
def AnonymizeWord(W, FirstNameR, LastNameR, FirstNameO, LastNameO, F):
    # make sure everything is in lowercase in local copies
    result = W.lower()
    _FirstNameR = FirstNameR.lower();
    _LastNameR = LastNameR.lower();
    _FirstNameO = FirstNameO.lower();
    _LastNameO = LastNameO.lower();

    # for each rule first check to see if the word has already changed, if it is already changed then do not run anymore rules
#    print(result)
    result = AnonymizeWord_Pronoun(result)
#    print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Adjective(result)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Abbreviations(result)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Name(result, _FirstNameR, _LastNameR)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Name(result, _FirstNameO, _LastNameO)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_NickName(result, _FirstNameR, _LastNameR)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_NickName(result, _FirstNameO, _LastNameO)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_NameSubword(result, _FirstNameR)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_NameSubword(result, _LastNameR)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_NameSubword(result, _FirstNameO)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_NameSubword(result, _LastNameO)
        #print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Flag(result, F)
        #print(result)

    return result
"""

"""
Inputs:
    String W: The current word under consideration
    Array[String] Names: The array of names to consider
    Boolean F: A flag that indicates the previous word was unusual

Returns:
    String: the word W or the anonymized word

AnonymizeWord functions by calling several other functions to do the actual work of anonymization.
Think of it as a collection of rules to be applied.
"""


def AnonymizeWord(W, Names, F):
    # make sure everything is in lowercase in local copies
    result = W.lower()

    # for each rule first check to see if the word has already changed, if it is already changed then do not run anymore rules
    #    print(result)
    result = AnonymizeWord_Pronoun(result)
    #    print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Adjective(result)
        # print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Abbreviations(result)
        # print(result)

    for name in Names:
        _name = name.lower();
        if (result == W.lower()):
            result = AnonymizeWord_Name(result, _name)
            # print(result)

        if (result == W.lower()):
            result = AnonymizeWord_NickName(result, _name)
            # print(result)

        if (result == W.lower()):
            result = AnonymizeWord_NameSubword(result, _name)
            # print(result)

    if (result == W.lower()):
        result = AnonymizeWord_Flag(result, F)
        # print(result)

    if (result == W.lower()):
        result = None

    return result


def SetAnonFlag(W):
    result = False
    processing = W.lower();

    if ((processing == "doctor") or
            (processing == "dr") or
            (processing == "mister") or
            (processing == "miss") or
            (processing == "mr") or
            (processing == "ms") or
            (processing == "mrs")):
        result = True

    return result


result = ""
flag = False

# test 1: basic functionality
testWord = "Word"
names = ["Jason","Bernard","Brent","Thoma"]
result = AnonymizeWord(testWord, names, False)
flag = SetAnonFlag(testWord)
print((result if result else "NONE") + " " + str(flag))

# test: spellcheck
testWord = "Kevin"
names = ["Jason","Bernard","Brent","Thoma"]
result = AnonymizeWord(testWord, names, False)
flag = SetAnonFlag(testWord)
print((result if result else "NONE") + " " + str(flag))

# test: pronoun
testWordPronoun = "he"
result = AnonymizeWord(testWordPronoun, names, False)
flag = SetAnonFlag(testWordPronoun)
print((result if result else "NONE") + " " + str(flag))

# test: adjective
testWordAdjective = "female"
result = AnonymizeWord(testWordAdjective, names, False)
flag = SetAnonFlag(testWordAdjective)
print((result if result else "NONE") + " " + str(flag))

# test: abbreviation
testWordAbbreviation = "f"
result = AnonymizeWord(testWordAbbreviation, names, False)
flag = SetAnonFlag(testWordAbbreviation)
print((result if result else "NONE") + " " + str(flag))

# test: name
testWordName = "Jason"
testResidentFName = "Jason"
result = AnonymizeWord(testWordName, names, False)
flag = SetAnonFlag(testWordName)
print((result if result else "NONE") + " " + str(flag))

# test: nickname
testWordNickname = "Richie"
names = ["Richard","Bernard","Brent","Thoma"]
result = AnonymizeWord(testWordNickname, names,
                       False)
flag = SetAnonFlag(testWordNickname)
print((result if result else "NONE") + " " + str(flag))

# test: flag
testWordFlag1 = "Doctor"
result = AnonymizeWord(testWordFlag1, names, False)
flag = SetAnonFlag(testWordFlag1)
print((result if result else "NONE") + " " + str(flag))
testWordFlag2 = "Bernard"
result = AnonymizeWord(testWordFlag2, names, flag)
flag = SetAnonFlag(testWordFlag2)
print((result if result else "NONE") + " " + str(flag))